consensus/polybft/state_test.go

```diff
@@ -0,0 +1,614 @@
+	state := newTestState(t)
+
+	commitmentMsg, err := createTestCommitmentMessage(1, 10)
+	assert.NoError(t, err)
+
+	hash := commitmentMsg.Message.Hash()
+
+	assert.NoError(t, err)
+
+	_, err = state.insertMessageVote(0, hash.String(), &MessageSignature{
+		From:      "NODE_1",
+		Signature: []byte{1, 2},
+	})
+	assert.NoError(t, err)
+
+	commitmentMsg, err = createTestCommitmentMessage(11, 10)
+	assert.NoError(t, err)
+
+	hash = commitmentMsg.Message.Hash()
+
+	assert.NoError(t, err)
+
+	_, err = state.insertMessageVote(2, hash.String(), &MessageSignature{
+		From:      "NODE_1",
+		Signature: []byte{2, 3},
+	})
+	assert.NoError(t, err)
+	_, err = state.insertMessageVote(2, hash.String(), &MessageSignature{
+		From:      "NODE_2",
+		Signature: []byte{4, 5},
+	})
+	assert.NoError(t, err)
+
+	votes, err := state.getMessageVotes(hash.String())
+	assert.NoError(t, err)
+	assert.Equal(t, 2, len(votes.Signatures))
+}
+
+func TestState_InsertVoteConcurrent(t *testing.T) {
+	state := newTestState(t)
+
+	epoch := uint64(1)
+	hash := "hash"
+
+	var wg sync.WaitGroup
+
+	for i := 0; i < 100; i++ {
+		wg.Add(1)
+
+		go func(i int) {
+			defer wg.Done()
+
+			_, _ = state.insertMessageVote(epoch, hash, &MessageSignature{
+				From:      pbft.NodeID(fmt.Sprintf("NODE_%d", i)),
+				Signature: []byte{1, 2},
+			})
+		}(i)
+	}
+
+	wg.Wait()
+
+	votes, err := state.getMessageVotes(hash)
+	assert.NoError(t, err)
+	assert.Len(t, votes.Signatures, 100)
+}
+
+func TestState_Insert_And_Cleanup(t *testing.T) {
+	state := newTestState(t)
+
+	for i := uint64(1); i < 1001; i++ {
+		hash := fmt.Sprintf("hash_%v", i)
+		_, _ = state.insertMessageVote(i, hash, &MessageSignature{
+			From:      "NODE_1",
+			Signature: []byte{1, 2},
+		})
+	}
+
+	// BucketN returns number of all buckets inside root bucket (including nested buckets) + the root itself
+	// Since we inserted 1000 message votes for 1000 epochs we expect to have 1000 buckets
+	assert.Equal(t, 1000, state.bucketStats(messageVotesBucket).KeyN)
+
+	err := state.cleanPreviousEpochsDataFromDB(1001)
+	assert.NoError(t, err)
+
+	assert.Equal(t, 0, state.bucketStats(messageVotesBucket).BucketN-1)
+
+	// there should be no votes for given epoch since we cleaned the db
+	votes, err := state.getMessageVotes("hash_1000")
+	assert.NoError(t, err)
+	assert.Nil(t, votes)
+
+	for i := uint64(1001); i < 2001; i++ {
+		hash := fmt.Sprintf("hash_%v", i)
+		_, _ = state.insertMessageVote(i, hash, &MessageSignature{
+			From:      "NODE_1",
+			Signature: []byte{1, 2},
+		})
+	}
+
+	assert.Equal(t, 1000, state.bucketStats(messageVotesBucket).KeyN)
+
+	votes, err = state.getMessageVotes("hash_2000")
+	assert.NoError(t, err)
+	assert.Len(t, votes.Signatures, 1)
+}
+
+func TestState_insertAndGetValidatorSnapshot(t *testing.T) {
+	epoch := uint64(1)
+
+	state := newTestState(t)
+
+	keys, err := bls.CreateRandomBlsKeys(3)
+
+	require.NoError(t, err)
+
+	snapshot := AccountSet{
```

Comment:
Didn't we have helper functions for this?

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/785/commits/e6ea05a82b32ac463830791a0ab9cf661f5fa87c